modules/search/backend_search_block_writer.go

```diff
@@ -0,0 +1,121 @@
+package search
+
+import (
+	"bytes"
+	"context"
+
+	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/google/uuid"
+	"github.com/grafana/tempo/pkg/tempofb"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/grafana/tempo/tempodb/encoding"
+	"github.com/grafana/tempo/tempodb/encoding/common"
+)
+
+// backendSearchBlockWriter is a DataWriter for search data. Instead of receiving bytes slices, it
+// receives search data objects and combintes them into a single FlatBuffer Builder and
+// flushes periodically, one page per flush.
+type backendSearchBlockWriter struct {
+	// input
+	blockID  uuid.UUID
+	tenantID string
+	w        backend.RawWriter
+
+	// vars
+	builder     *flatbuffers.Builder
+	pageEntries []flatbuffers.UOffsetT
+	pageBuf     []byte
+	tracker     backend.AppendTracker
+	finalBuf    *bytes.Buffer
+	dw          common.DataWriter
+}
+
+var _ common.DataWriterGeneric = (*backendSearchBlockWriter)(nil)
+
+func newBackendSearchBlockWriter(blockID uuid.UUID, tenantID string, w backend.RawWriter, enc backend.Encoding) (*backendSearchBlockWriter, error) {
+	finalBuf := &bytes.Buffer{}
+
+	dw, err := encoding.LatestEncoding().NewDataWriter(finalBuf, enc)
+	if err != nil {
+		return nil, err
+	}
+
+	return &backendSearchBlockWriter{
+		blockID:  blockID,
+		tenantID: tenantID,
+		w:        w,
+
+		pageBuf:  make([]byte, 0, 1024*1024),
+		finalBuf: finalBuf,
+		builder:  flatbuffers.NewBuilder(1024),
+		dw:       dw,
+	}, nil
+}
+
+// Write the data to the flatbuffer builder. Input must be a SearchDataMutable. Returns
+// the number of bytes written, which is determined from the current object in the builder.
+func (w *backendSearchBlockWriter) Write(ctx context.Context, ID common.ID, i interface{}) (int, error) {
```

Comment:
Nice. This is a clever integration of the existing Writer interfaces and the flatbuffer paging.

I wonder if there's something we can do to reduce the complexity of all these Writers/Readers. Can we make the existing things generic enough to support this case as well as the current one? or would we just make everything more complex?

Even if we never merge this code with the existing set of writers/readers I think we need to come up with a code organization structure that works going forward. Should we push all this functionality into tempodb? I honestly don't know. Let's discuss.

Reply:
> Can we make the existing things generic enough to support this case as well as the current one? or would we just make everything more complex?

Right, was considering if AppenderBuffered should be a byte-oriented shim around the new generic one (just one small part). Agree, let's discuss.

Reply:
Tried out making AppenderBuffered a byte-oriented shim and it didn't seem worth the additional complexity. Will leave the implementations separate for now.